src/ripple/app/main/GRPCServer.h

```diff
@@ -0,0 +1,349 @@
+
+
+        public:
+
+        virtual ~CallData() {}
+  
+        // Take in the "service" instance (in this case representing an
+        // asynchronous server) and the completion queue "cq" used for
+        // asynchronous communication with the gRPC runtime.
+        CallData(
+                rpc::v1::XRPLedgerAPIService::AsyncService& service,
+                grpc::ServerCompletionQueue& cq,
+                Application& app,
+                BindListener<Request,Response> bind_listener,
+                Handler<Request,Response> handler,
+                RPC::Condition required_condition,
+                Resource::Charge load_type);
+   
+  
+        void process() override;
+  
+        bool isFinished() override
+        {
+            //checking the status while a request is in the middle of being
+            //processed will lead to indeterminate results. Lock here to 
+            //sequence checking status and processing
+            std::lock_guard<std::mutex> lock(mut_);
+            return status_ == CallStatus::FINISH;
+        }
+  
+        virtual void abort() override
+        {
+            std::lock_guard<std::mutex> lock(mut_);
+            aborted_ = true;
+        }
+  
+        void set_iter(std::list<std::shared_ptr<Processor>>::iterator const & it) override
+        {
+            iter_ = it;
+        }
+  
+        std::list<std::shared_ptr<Processor>>::iterator get_iter() override
+        {
+            if(!iter_)
+                BOOST_ASSERT(false);
+  
+            return iter_.get();
+        }
+  
+        std::shared_ptr<Processor> clone() override;
+  
+        private:
+  
+        //process the request. Called inside the coroutine passed to JobQueue
+        void process(std::shared_ptr<JobQueue::Coro> coro);
+  
+        //return load type of this RPC
+        Resource::Charge getLoadType()
+        {
+            return load_type_;
+        }
+  
+        //return the Role required for this RPC
+        //for now, we are only supporting RPC's that require Role::USER for gRPC
+        Role getRole()
+        {
+            return Role::USER;
+        }
+  
+        //register endpoint with ResourceManager and return usage
+        Resource::Consumer getUsage()
+        {
+            std::string peer = getEndpoint(ctx_.peer());
+            boost::optional<beast::IP::Endpoint> endpoint =
+                beast::IP::Endpoint::from_string_checked(peer);
+            return app_.getResourceManager().newInboundEndpoint(endpoint.get());
+        }
+
+    }; //CallData
+  
+  
+    //Create a CallData object for each RPC
+    void setupListeners();
+  
+    //make a CallData instance, returned as shared_ptr to base class (Processor)
+    template <class Request, class Response>
+    std::shared_ptr<Processor> makeCallData(
+            BindListener<Request,Response> bl,
+            Handler<Request,Response> handler,
+            RPC::Condition condition,
+            Resource::Charge load_type)
```

Comment:
I'd remove this function and just call `make_shared` directly where needed.

Reply:
removed